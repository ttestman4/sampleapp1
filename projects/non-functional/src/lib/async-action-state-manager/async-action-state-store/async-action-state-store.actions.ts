import { Action } from '@ngrx/store';

export enum AsyncStateActionTypes {
    Start = '[Async Action State] Start',
    Success = '[Async Action State] Success',
    Error = '[Async Action State] Error',
}

/**
 * Every action is comprised of at least a type and an optional
 * payload. Expressing actions as classes enables powerful
 * type checking in reducer functions.
 *
 * See Discriminated Unions: https://www.typescriptlang.org/docs/handbook/advanced-types.html#discriminated-unions
 */
export class Start implements Action {
    readonly type = AsyncStateActionTypes.Start;
    constructor(public payload: string) { }
}

export class Success implements Action {
    readonly type = AsyncStateActionTypes.Success;
    constructor(public payload: string) { }
}

export class Error implements Action {
    readonly type = AsyncStateActionTypes.Error;
    constructor(public payload: string) { }
}

export type AsyncStateActionsUnion =
    Start |
    Success |
    Error;
