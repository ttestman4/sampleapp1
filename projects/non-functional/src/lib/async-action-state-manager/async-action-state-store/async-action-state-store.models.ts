import { Action } from '@ngrx/store';

export enum AsyncStatus {
    Off = '[Async Status] Off',
    Start = '[Async Status] Start',
    Success = '[Async Status] Success',
    Error = '[Async Status] Error',
}

export interface AsyncActionState {
    asyncActionId: string;
    status: AsyncStatus;
}

export interface AsyncAction extends Action {
    asyncActionState: AsyncActionState;
}
