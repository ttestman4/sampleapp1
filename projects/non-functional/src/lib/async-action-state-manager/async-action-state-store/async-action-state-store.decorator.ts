import * as AsyncStateModels from './async-action-state-store.models';

export function AsyncActionStart(actionId: string) {
    return startClassDecorator;

    function startClassDecorator<T extends { new(...args: any[]): {} }>(constructor: T) {
        return class extends constructor {
            asyncActionState: AsyncStateModels.AsyncActionState = {
                asyncActionId: actionId,
                status: AsyncStateModels.AsyncStatus.Start,
            };
        };
    }
}

export function AsyncActionSuccess(actionId: string) {
    return successClassDecorator;

    function successClassDecorator<T extends { new(...args: any[]): {} }>(constructor: T) {
        return class extends constructor {
            asyncActionState: AsyncStateModels.AsyncActionState = {
                asyncActionId: actionId,
                status: AsyncStateModels.AsyncStatus.Success,
            };
        };
    }
}

export function AsyncActionError(actionId: string) {
    return errorClassDecorator;

    function errorClassDecorator<T extends { new(...args: any[]): {} }>(constructor: T) {
        return class extends constructor {
            asyncActionState: AsyncStateModels.AsyncActionState = {
                asyncActionId: actionId,
                status: AsyncStateModels.AsyncStatus.Error,
            };
        };
    }
}
