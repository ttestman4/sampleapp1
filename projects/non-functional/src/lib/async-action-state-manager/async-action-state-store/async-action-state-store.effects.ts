import { Injectable } from '@angular/core';
import { Actions, Effect } from '@ngrx/effects';
import { filter, map } from 'rxjs/operators';
import * as util from '../../helper-functions/helper-functions';
import * as AsyncStateActions from './async-action-state-store.actions';
import * as AsyncStateModels from './async-action-state-store.models';

let nextAsyncActionEffectsId = 1;

export function ResetNextAsyncActionEffectsId() {
    nextAsyncActionEffectsId = 1;
}
@Injectable({
    providedIn: 'root'
})
export class AsyncActionStateEffects {

    @Effect()
    asyncStateUpdater$ = this.actions$.pipe(
        filter((action: AsyncStateModels.AsyncAction) =>
            (util.isObject(action.asyncActionState) &&
                util.isString(action.asyncActionState.asyncActionId) &&
                (action.asyncActionState.status === AsyncStateModels.AsyncStatus.Start ||
                    action.asyncActionState.status === AsyncStateModels.AsyncStatus.Success ||
                    action.asyncActionState.status === AsyncStateModels.AsyncStatus.Error))
        ),
        map(action => {
            switch (action.asyncActionState.status) {
                case AsyncStateModels.AsyncStatus.Start: {
                    return new AsyncStateActions.Start(action.asyncActionState.asyncActionId);
                }
                case AsyncStateModels.AsyncStatus.Success: {
                    return new AsyncStateActions.Success(action.asyncActionState.asyncActionId);
                }
                case AsyncStateModels.AsyncStatus.Error: {
                    return new AsyncStateActions.Error(action.asyncActionState.asyncActionId);
                }
                default: {
                    return new AsyncStateActions.Error(action.asyncActionState.asyncActionId);
                }
            }
        })
    );

    constructor(
        private actions$: Actions<AsyncStateModels.AsyncAction>
    ) {
        const id = nextAsyncActionEffectsId++;
        if (id > 1) {
            throw new Error(
                'AsyncActionStateEffects is already loaded. Import it in the AppModule only');
        }
    }
}
