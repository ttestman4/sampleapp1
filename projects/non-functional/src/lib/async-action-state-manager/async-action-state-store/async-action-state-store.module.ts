import { ModuleWithProviders, NgModule, Optional, SkipSelf  } from '@angular/core';
import { CommonModule } from '@angular/common';
import { EffectsModule } from '@ngrx/effects';
import { StoreModule } from '@ngrx/store';
import { featureName, reducer } from './async-action-state-store.reducer';
import { ResetNextAsyncActionEffectsId, AsyncActionStateEffects } from './async-action-state-store.effects';

export * from './async-action-state-store.actions';
export * from './async-action-state-store.models';
export * from './async-action-state-store.decorator';
export * from './async-action-state-store.selectors';
@NgModule({
  declarations: [],
  imports: [
    CommonModule,
    /**
     * StoreModule.forFeature is used for composing state
     * from feature modules. These modules can be loaded
     * eagerly or lazily and will be dynamically added to
     * the existing state.
     */
    StoreModule.forFeature(featureName, reducer),
    /**
     * Effects.forFeature is used to register effects
     * from feature modules. Effects can be loaded
     * eagerly or lazily and will be started immediately.
     *
     * All Effects will only be instantiated once regardless of
     * whether they are registered once or multiple times.
     */
    EffectsModule.forFeature([AsyncActionStateEffects]),
  ]
})
export class AsyncActionStateStoreModule {
  constructor(@Optional() @SkipSelf() parentModule: AsyncActionStateStoreModule) {
    if (parentModule) {
      throw new Error(
        'AsyncActionStateStoreModule is already loaded. Import it in the AppModule only');
    }
  }

  static forRoot(): ModuleWithProviders {
    return {
      ngModule: AsyncActionStateStoreModule
    };
  }
  static forChild(): ModuleWithProviders {
    return {
      ngModule: AsyncActionStateStoreModule,
    };
  }
  static forTestReset() {
    ResetNextAsyncActionEffectsId();
  }
}
