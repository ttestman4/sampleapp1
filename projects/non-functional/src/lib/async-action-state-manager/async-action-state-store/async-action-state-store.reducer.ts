import { createEntityAdapter, EntityAdapter, EntityState } from '@ngrx/entity';
import { createFeatureSelector } from '@ngrx/store';
import * as AsyncStateActions from './async-action-state-store.actions';
import * as AsyncStateModels from './async-action-state-store.models';

export const featureName = 'asyncState';

export interface State extends EntityState<AsyncStateModels.AsyncActionState> {
    name: string;
}

export const adapter: EntityAdapter<AsyncStateModels.AsyncActionState> = createEntityAdapter<AsyncStateModels.AsyncActionState>({
    selectId: (a) => a.asyncActionId,
    sortComparer: false,
});

export const initialState: State = adapter.getInitialState({
    name: 'Async Action State Store',
});

export function reducer(
    state = initialState,
    action: AsyncStateActions.AsyncStateActionsUnion
): State {
    switch (action.type) {
        case AsyncStateActions.AsyncStateActionTypes.Start: {
            return adapter.upsertOne({
                asyncActionId: action.payload,
                status: AsyncStateModels.AsyncStatus.Start,
            }, state);
        }
        case AsyncStateActions.AsyncStateActionTypes.Success: {
            return adapter.upsertOne({
                asyncActionId: action.payload,
                status: AsyncStateModels.AsyncStatus.Success,
            }, state);
        }
        case AsyncStateActions.AsyncStateActionTypes.Error: {
            return adapter.upsertOne({
                asyncActionId: action.payload,
                status: AsyncStateModels.AsyncStatus.Error,
            }, state);
        }
        default: {
            return state;
        }
    }
}

export const selectAsyncActionState = createFeatureSelector<State>(featureName);

export const {
    selectEntities: selectAsyncStateEntitiesFromState,
} = adapter.getSelectors();
