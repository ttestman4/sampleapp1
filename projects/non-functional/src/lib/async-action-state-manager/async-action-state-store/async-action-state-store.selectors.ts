import { createSelector } from '@ngrx/store';
import { selectAsyncActionState, selectAsyncStateEntitiesFromState } from './async-action-state-store.reducer';

export const selectAsyncStateEntities = createSelector(selectAsyncActionState,
    selectAsyncStateEntitiesFromState
);
