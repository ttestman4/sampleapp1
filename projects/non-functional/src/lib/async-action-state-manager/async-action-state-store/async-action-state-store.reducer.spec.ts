import { initialState, reducer, State } from './async-action-state-store.reducer';
import * as AsyncStateActions from './async-action-state-store.actions';
import * as AsyncStateModels from './async-action-state-store.models';
describe('Async State Reducer', () => {
    it('Unknown action should return the initial state', () => {
        const action = {} as any;

        const result = reducer(initialState, action);
        expect(result).toEqual(initialState);
        expect(result).toMatchSnapshot();
    });

    it('Unknown action should return the initial state for undefined initial state input', () => {
        const action = {} as any;

        const result = reducer(undefined, action);
        expect(result).toEqual(initialState);
        expect(result).toMatchSnapshot();
    });

    let resultFromPreviousStep: State;

    describe('Start', () => {
        it('Valid action should return the added Async Action Id', () => {
            const action = new AsyncStateActions.Start('foo');

            resultFromPreviousStep = reducer(initialState, action);

            expect(resultFromPreviousStep.entities['foo'].status)
                .toEqual(AsyncStateModels.AsyncStatus.Start);
            expect(resultFromPreviousStep).toMatchSnapshot();
        });
    });

    describe('Success', () => {
        it('Valid action should return the added Async Action Id', () => {
            const action = new AsyncStateActions.Success('foo');

            resultFromPreviousStep = reducer(initialState, action);

            expect(resultFromPreviousStep.entities['foo'].status)
                .toEqual(AsyncStateModels.AsyncStatus.Success);
            expect(resultFromPreviousStep).toMatchSnapshot();
        });
    });

    describe('Error', () => {
        it('Valid action should return the added Async Action Id', () => {
            const action = new AsyncStateActions.Error('foo');

            resultFromPreviousStep = reducer(initialState, action);

            expect(resultFromPreviousStep.entities['foo'].status)
                .toEqual(AsyncStateModels.AsyncStatus.Error);
            expect(resultFromPreviousStep).toMatchSnapshot();
        });
    });
});
