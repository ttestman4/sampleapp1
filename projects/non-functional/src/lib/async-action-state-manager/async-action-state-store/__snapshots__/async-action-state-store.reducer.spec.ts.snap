// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Async State Reducer Error Valid action should return the added Async Action Id 1`] = `
Object {
  "entities": Object {
    "foo": Object {
      "asyncActionId": "foo",
      "status": "[Async Status] Error",
    },
  },
  "ids": Array [
    "foo",
  ],
  "name": "Async Action State Store",
}
`;

exports[`Async State Reducer Start Valid action should return the added Async Action Id 1`] = `
Object {
  "entities": Object {
    "foo": Object {
      "asyncActionId": "foo",
      "status": "[Async Status] Start",
    },
  },
  "ids": Array [
    "foo",
  ],
  "name": "Async Action State Store",
}
`;

exports[`Async State Reducer Success Valid action should return the added Async Action Id 1`] = `
Object {
  "entities": Object {
    "foo": Object {
      "asyncActionId": "foo",
      "status": "[Async Status] Success",
    },
  },
  "ids": Array [
    "foo",
  ],
  "name": "Async Action State Store",
}
`;

exports[`Async State Reducer Unknown action should return the initial state 1`] = `
Object {
  "entities": Object {},
  "ids": Array [],
  "name": "Async Action State Store",
}
`;

exports[`Async State Reducer Unknown action should return the initial state for undefined initial state input 1`] = `
Object {
  "entities": Object {},
  "ids": Array [],
  "name": "Async Action State Store",
}
`;
